.github/workflows/ci.yml

```diff
@@ -11,209 +11,239 @@ concurrency:
-        with:
-          directory: empty
-          token: 87cefb17-c44b-4f2f-8b30-1fff5769ce46
-          name: Ubuntu (${{ matrix.python }}${{ matrix.extra_name }})
-          flags: Ubuntu,${{ matrix.python }}
-
-  macOS:
-    name: 'macOS (${{ matrix.python }})'
-    timeout-minutes: 15
-    runs-on: 'macos-latest'
+  Cython:
+    name: "Cython"
+    runs-on: ubuntu-latest
     strategy:
       fail-fast: false
       matrix:
-        python: ['3.8', '3.9', '3.10', 'pypy-3.9-nightly', 'pypy-3.10-nightly']
-    continue-on-error: >-
-      ${{
-        (
-          endsWith(matrix.python, '-dev')
-          || endsWith(matrix.python, '-nightly')
-        )
-        && true
-        || false
-      }}
+        cython: ['==0.29.23', '==0.29.26', '==0.29.30', '<3', '>=3']
+        python: ['3.8', '3.10', '3.11', '3']
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
       - name: Setup python
-        uses: actions/setup-python@v4
-        with:
-          python-version: ${{ fromJSON(format('["{0}", "{1}"]', format('{0}.0-alpha - {0}.X', matrix.python), matrix.python))[startsWith(matrix.python, 'pypy')] }}
-          cache: pip
-          cache-dependency-path: test-requirements.txt
-      - name: Run tests
-        run: ./ci.sh
-      - if: always()
-        uses: codecov/codecov-action@v3
-        with:
-          directory: empty
-          token: 87cefb17-c44b-4f2f-8b30-1fff5769ce46
-          name: macOS (${{ matrix.python }})
-          flags: macOS,${{ matrix.python }}
-
-  # run CI on a musl linux
-  Alpine:
-    name: "Alpine"
-    runs-on: ubuntu-latest
-    container: alpine
-    steps:
-      - name: Checkout
-        uses: actions/checkout@v3
-      - name: Install necessary packages
-        # can't use setup-python because that python doesn't seem to work;
-        # `python3-dev` (rather than `python:alpine`) for some ctypes reason,
-        # `nodejs` for pyright (`node-env` pulls in nodejs but that takes a while and can time out the test).
-        run: apk update && apk add python3-dev bash nodejs
-      - name: Enter virtual environment
-        run: python -m venv .venv
-      - name: Run tests
-        run: source .venv/bin/activate && ./ci.sh
-      - if: always()
-        uses: codecov/codecov-action@v3
+        uses: actions/setup-python@v5
         with:
-          directory: empty
-          token: 87cefb17-c44b-4f2f-8b30-1fff5769ce46
-          name: Alpine
-          flags: Alpine,3.12
+            python-version: '${{ matrix.python }}'
+            cache: pip
+      - name: install cython
+        run: |
+          # setuptools is needed to get distutils on 3.12
+          python -m pip install --upgrade pip setuptools
+          python -m pip install "cython${{ matrix.cython}}"
+      - name: install trio
+          # use -e?
+        run: python -m pip install .
+      - name: compile pyx file
+        run: cythonize -i tests/cython/test_cython.pyx
```

Comment:
It should be fine to run the compiled file across multiple python versions -- maybe that'll make for a faster CI? (probably just a future optimization tbh)

Reply:
looking at the tests, that step is taking ~ 1-3s. So negligible time save, and all but two of the tests are failing on the compile step (though due to super old cython + new python version, which maybe isn't terribly relevant for us).

Reply:
What would likely save a significant amount of time is only having separate machines for the python versions, and testing all relevant cython versions on the same machine. Since the job itself is so short the gain from parallelization is vastly outweighed by having to wait for acquisition of the machine and setup/teardown.